### PR TITLE
Prototype gradient descent method with least squares

### PR DESCRIPTION
**Pull request by aeslaughter**
@zachmprince What do you think of this concept for the start of ML based optimization? If you think this is alright I can make this work with @grmnptr effort to have two methods that can train a PolynomialModel object.

@cbolisetti You might also be interested in this, perhaps you can work on this if you want to get your feet wet with STM? 

(refs #15361)

### CODE REVIEW

File: modules/stochastic_tools/src/surrogates/GradientDescentTrainer.C [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        _A_matrix(row, col) = std::pow(_x_training_data[row], col);
+  }
+
+  // compute A^T(Ax-b)
+  _A_matrix.vector_mult(_Ax_minus_b, x);
+  _Ax_minus_b -= _b_vector;
+
+  DenseVector<Real> x_out;
+  _A_matrix.vector_mult_transpose(x_out, _Ax_minus_b);
+  return x_out;
```
Comment (aeslaughter) [inline modules/stochastic_tools/src/surrogates/GradientDescentTrainer.C:94]: @fdkong Can PETSc do this sort of computation with a distributed dense matrix (https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Mat/MatCreateDense.html)?

Comment (fdkong) [reply modules/stochastic_tools/src/surrogates/GradientDescentTrainer.C:94]: I will be surprised if it can not. But I did not try such things so far.

MatCreateDense

MatMultTranspose

**Comment by aeslaughter** [comment]
I also think it should be possible so separate the LeastSquares from the Polynomial, such that we can inject other functions that can be described as Ax, such as a Fourier series.

**Comment by aeslaughter** [comment]
I am also going to separate the LeastSquares from the Polynomial and add a Fourier series example as well.

**Comment by cbolisetti** [comment]
> @zachmprince What do you think of this concept for the start of ML based optimization? If you think this is alright I can make this work with @grmnptr effort to have two methods that can train a PolynomialModel object.
> 
> @cbolisetti You might also be interested in this, perhaps you can work on this if you want to get your feet wet with STM?
> 
> (refs #15361)

@aeslaughter Yes, I'll be happy to!

**Comment by aeslaughter** [comment]
I started to create a Polynomial sampler that will generate some fuzzy data along a line, but it uncovered some sort of parallel data bug in Sampler and/or SamplerData. I will figure that out and then go back to working on SGD in parallel, with distributed data. @cbolisetti When I get this parallel stuff figured out, I will have you start working on this.

> I am also going to separate the LeastSquares from the Polynomial and add a Fourier series example as well.

To reply to myself, I don't think trying to separate these further is worth the time, trying to over generalize it seems to make the whole thing too complex. I think we should create an Optimizer object that inherits from Trainer, all an Optimizer adds is the requirement of an ObjectiveFunction (and in the future an optional Regularizer). We can than create various Optimizer and ObjectiveFunction objects that should work interchangeably.
